controller/cmd/service-mirror/cluster_watcher.go

```diff
@@ -0,0 +1,782 @@
+		ns := &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: map[string]string{
+					consts.MirroredResourceLabel:  "true",
+					consts.RemoteClusterNameLabel: rcsw.clusterName,
+				},
+				Name: namespace,
+			},
+		}
+		_, err := rcsw.localAPIClient.Client.CoreV1().Namespaces().Create(ns)
+		if err != nil {
+			return fmt.Errorf("could not create mirror namespace: %s", err)
+		}
+	}
+	return nil
+}
+
+// This method takes care of port remapping. What it does essentially is get the one gateway port
+// that we should send traffic to and create endpoint ports that bind to the mirrored service ports
+// (same name, etc) but send traffic to the gateway port. This way we do not need to do any remapping
+// on the service side of things. It all happens in the endpoints.
+func (rcsw *RemoteClusterServiceWatcher) getEndpointsPorts(service *corev1.Service, gatewayPort int32) []corev1.EndpointPort {
+	var endpointsPorts []corev1.EndpointPort
+	for _, remotePort := range service.Spec.Ports {
+		endpointsPorts = append(endpointsPorts, corev1.EndpointPort{
+			Name:     remotePort.Name,
+			Protocol: remotePort.Protocol,
+			Port:     gatewayPort,
+		})
+	}
+	return endpointsPorts
+}
+
+func (rcsw *RemoteClusterServiceWatcher) cleanupOrphanedServices() error {
+	matchLabels := map[string]string{
+		consts.MirroredResourceLabel:  "true",
+		consts.RemoteClusterNameLabel: rcsw.clusterName,
+	}
+
+	servicesOnLocalCluster, err := rcsw.listLocalServices(matchLabels)
+	if err != nil {
+		return fmt.Errorf("failed obtaining local services while GC-ing: %s", err)
+	}
+
+	for _, srv := range servicesOnLocalCluster {
+		_, err := rcsw.remoteAPIClient.Svc().Lister().Services(srv.Namespace).Get(rcsw.originalResourceName(srv.Name))
+		if err != nil {
+			// service does not exist anymore. Need to delete
+			if err := rcsw.localAPIClient.Client.CoreV1().Services(srv.Namespace).Delete(srv.Name, &metav1.DeleteOptions{}); err != nil {
+				rcsw.log.Errorf("Failed to GC local service %s", srv.Name)
+			} else {
+				rcsw.log.Debugf("Deleted service %s/%s as part of GC process", srv.Namespace, srv.Name)
+			}
+		}
+	}
+	return nil
+}
+
+// Whenever we stop watching a cluster, we need to cleanup everything that we have
+// created. This piece of code is responsible for doing just that. It takes care of
+// services, endpoints and namespaces (if needed)
+func (rcsw *RemoteClusterServiceWatcher) cleanupMirroredResources() error {
+	matchLabels := map[string]string{
+		consts.MirroredResourceLabel:  "true",
+		consts.RemoteClusterNameLabel: rcsw.clusterName,
+	}
+
+	services, err := rcsw.listLocalServices(matchLabels)
+	if err != nil {
+		return fmt.Errorf("could not retrieve mirrored services that need cleaning up: %s", err)
+	}
+
+	for _, svc := range services {
+		if err := rcsw.localAPIClient.Client.CoreV1().Services(svc.Namespace).Delete(svc.Name, &metav1.DeleteOptions{}); err != nil {
```

Comment:
So we cannot delete namespaces if a ns is left empty, because we dont know whether its empt or not. If we are stopping to watch the cluster though, we can delete the NS. SHould we unify it and make sure we never delete namespaces ?

Reply:
Yeah, I think never deleting namespaces is the right call.

Reply:
Just so I understand, is the reasoning behind never deleting namespaces because we could have created a mirror service in a Namespace of our current cluster which previously existed and has non-mirrored resources and we don't want to accidentally delete those?